tibanna/ec2_utils.py

```diff
@@ -738,65 +710,175 @@ def create_userdata(self, profile=None):
-        if self.cfg.availability_zone:
-            largs.update({'Placement': {'AvailabilityZone': self.cfg.availability_zone}})
-        if self.cfg.security_group:
-            largs.update({'SecurityGroupIds': [self.cfg.security_group]})
+            launch_template_data.update(
+                {'InstanceMarketOptions': 
+                    {'MarketType': 'spot','SpotOptions': spot_options}
+                }
+            )
+
+        try:
+            ec2.create_launch_template(
+                DryRun=self.dryrun,
+                LaunchTemplateName=launch_template_name,
+                LaunchTemplateData=launch_template_data,
+            )
+        except Exception as e:
+            raise Exception(f"Could not create launch template: {str(e)}")
+
+
+    def create_fleet(self):
+        '''Create an 'instant' type fleet with 1 instance'''
+        self.create_launch_template()
+
+        ec2 = boto3.client('ec2')
+        try:
+            fleet_spec = self.create_fleet_spec()
+            fleet_result = ec2.create_fleet(**fleet_spec)
+            return fleet_result
+        except Exception as e:
+            raise Exception(f"Unable to create fleet: {str(e)}")
+
+
+    def create_fleet_spec(self): # Factored out for easier testing
+
+        potential_ec2s = [] # Used as overrides in the launch template
+
+        subnets = False
         if self.cfg.subnet:
             if isinstance(self.cfg.subnet, str):
-                largs.update({'SubnetId': self.cfg.subnet})
+                subnets = [self.cfg.subnet]
             elif isinstance(self.cfg.subnet, list):
-                largs.update({'SubnetId': choice(self.cfg.subnet)})  # if given multiple, pick one randomly
-        if self.dryrun:
-            largs.update({'DryRun': True})
-        return largs
+                subnets = self.cfg.subnet
+
+        # Create all possible combinations of instance type / subnet
+        for instance_type in self.instance_type_list:
+            instance_info = self.instance_type_infos[instance_type]
+            if subnets:
+                for subnet in subnets:
+                    potential_ec2s.append({
+                        "InstanceType": instance_type,
+                        "SubnetId": subnet,
+                        "ImageId": instance_info['ami_id']
+                    })
+            else:
+                potential_ec2s.append({
+                    "InstanceType": instance_type,
+                    "ImageId": instance_info['ami_id']
+                })
+        
+        spec = {
+            "DryRun": self.dryrun,
+            "SpotOptions": {
+                'AllocationStrategy': 'capacity-optimized',
+                'InstanceInterruptionBehavior': 'terminate' # hibernate is an option here
+            },
+            "OnDemandOptions": {
+                'AllocationStrategy': 'lowest-price' # prioritized is the other option, but then we need to assign prioritoes in the launch template
```

Comment:
I think `lowest-price` is totally reasonable, no need to push more complexity onto the user. However, if we specify smaller machines that will get the job done slower in the instance of potentially working on spot it may be smart to eventually implement a priority strategy.

Reply:
Agreed, but I will leave it as is for now.